rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+
+#### Configuration model
+Beside the attributes that are defined by the provider for an ephemeral resource, the following meta-arguments needs to be supported by any ephemeral block:
+* lifecycle
+* count
+* for_each
+* depends_on
+* provider
+
+#### `Open` method details
+When OpenTofu will have to use an ephemeral resource, it needs to call its `Open` method, passing over the config of the ephemeral resource.
+
+The call to the `Open` method will return the following data:
+* `Private` that OpenTofu is not going to use in other contexts than calling the provider `Close` or `Renew` optionally defined methods.
+* `Result` that will contain the actual ephemeral information. This is what OpenTofu needs to handle to make it available to other ephemeral contexts to reference.
+* `RenewAt` being an optional timestamp indicating when OpenTofu will have to call `Renew` method on the provider before using again the data from the `Result`.
+
+Observations:
+* In the `Result`, OpenTofu is epecting to find any non-computed given values in the request, otherwise will return an error.
+* In the `Result`, the fields marked as computed can be either null or have an actual value. If an unknown if found, OpenTofu will return an error.
+
+> [!NOTE]
+>
+> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.
+
+#### `Renew` method details
+The `Renew` method is called only if the response from `Open` or another `Renew` call is containing a `RenewAt`.
+When `RenewAt` is present, OpenTofu, before using the `Result` from the `Open` method response, will check if the current timestamp is at or over `RenewAt` and will call the `Renew` method by providing the previously returned `Private` information.
+
+> [!NOTE]
+> 
+> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
+The `Renew` request is requiring the latest `Private` data returned by the call to `Open` or `Renew` method.
+
+#### `ConfigValidators` and `ValidateConfig` methods details
+There is not much to say here, since this is the same lifecycle that a datasource is having.
+
+#### Testing support
+The testing support will be documented later into a different RFC, or as amendment to this one.
```

Comment:
Can you give a reason as to why we're not putting it in this RFC?

Reply:
I didn't dig into this part yet and I barely touched the testing feature of this project.
If you think that this should be added, I can try my best but might take some time until I get a good understanding of the things in there.

Reply:
Given the limited usage of the testing feature, I agree with @yottta here.  We should focus on the main impact of this feature.

Reply:
I think I'm just looking for some reasoning to be added to the RFC. I'm fine with not doing it now, we should just document why.

Reply:
Quick fix: dd4ca4729bfb730c406cd06ceaa52e2d6fcfea25